backend/src/apiserver/server/api_util.go

```diff
@@ -137,14 +161,57 @@ func GetResourceReferenceFromRunInterface(r interface{}) ([]*apiv1beta1.Resource
 	}
 }
 
-// Converts a yaml string into structpb.Struct (v2).
-func yamlStringToProtobufStruct(s string) (*structpb.Struct, error) {
-	st := &structpb.Struct{}
-	err := yaml.Unmarshal([]byte(s), st)
-	if err != nil {
-		return nil, util.Wrap(err, "Failed to convert a yaml string into a protobuf struct")
+// Converts pipeline spec in yaml string into structpb.Struct (v2).
+// The string may contain multiple yaml documents, where there has to be one doc for pipeline spec,
+// and there may be a second doc for platform specific spec.
+// If platform spec is empty, then return pipeline spec directly. Else, return a struct with
+// pipeline spec and platform spec as subfields.
+func yamlStringToPipelineSpecStruct(s string) (*structpb.Struct, error) {
+	if s == "" {
+		return nil, nil
+	}
+	var pipelineSpec structpb.Struct
+	var platformSpec structpb.Struct
+	hasPipelineSpec := false
+	hasPlatformSpec := false
+
+	yamlStrings := strings.Split(s, "\n---\n")
+	for _, yamlString := range yamlStrings {
+		if template.IsPlatformSpecWithKubernetesConfig([]byte(yamlString)) {
+			hasPlatformSpec = true
+			jsonBytes, err := yaml.YAMLToJSON([]byte(yamlString))
+			if err != nil {
+				return nil, util.Wrap(err, "Failed to convert platformSpec yaml string into a protobuf struct")
+			}
+			err = protojson.Unmarshal(jsonBytes, &platformSpec)
+			if err != nil {
+				return nil, util.Wrap(err, "Failed to convert platformSpec yaml string into a protobuf struct")
+			}
+		} else {
```

Comment:
Should we also call `IsPipelineSpec` here? If `IsPlatformSpecWithKubernetesConfig` returns `false` instead of `true`, we will return error from `err = protojson.Unmarshal(jsonBytes, &pipelineSpec)`, which may be less convenient to debug for a user. WDYT?

Reply:
This was what I did initially, but we had test cases with minimum pipeline spec, which means we didn't check the validity of pipeline spec here. I'm not sure if we should have a behavior change here.